lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -56,14 +56,204 @@ removeNode = (node) ->
+    (orderChangePair.filter (link) -> link.href != linkToMove.href)[0]
+
+  [linkToMove, linksToPassBy]
+
+removeOrderChange = (link1, link2, orderChanges) ->
+  index = orderChanges.indexOf((orderChanges.filter (orderChangePair) ->
+    link1 in orderChangePair && link2 in orderChangePair)[0])
+  orderChanges.splice(index, 1)
+
+addOrderChange = (link1, link2, orderChanges) ->
+  orderChanges.push [link1, link2]
+
+reorderActiveLinks = (activeLinks, upstreamLinks) ->
+  orderedActiveLinks = activeLinks.slice()
+  orderChangesBetweenLinksToPerform = generateOrderChangesGraph(orderedActiveLinks, upstreamLinks)
+
+  while orderChangesBetweenLinksToPerform.length > 0
+    [linkToMove, linksToPassBy] = nextLinkToMove(orderedActiveLinks, orderChangesBetweenLinksToPerform)
+
+    moveDirection = if orderedActiveLinks.indexOf(linkToMove) > orderedActiveLinks.indexOf(linksToPassBy[0]) then 'UP' else 'DOWN'
+    startIndex = orderedActiveLinks.indexOf(linkToMove)
+
+    switch moveDirection
+      when 'UP'
+        for i in [(startIndex - 1)...0]
+          if orderedActiveLinks[i] in linksToPassBy
+            removeOrderChange(linkToMove, orderedActiveLinks[i], orderChangesBetweenLinksToPerform)
+            linksToPassBy.splice(linksToPassBy.indexOf(orderedActiveLinks[i]), 1)
+            if linksToPassBy.length == 0
+              document.head.removeChild(linkToMove)
+              triggerEvent('page:after-link-removed', linkToMove)
+              orderedActiveLinks.splice(startIndex, 1)
+
+              document.head.insertBefore(linkToMove, orderedActiveLinks[i])
+              orderedActiveLinks.splice(i, 0, linkToMove)
+              triggerEvent('page:after-link-inserted', linkToMove)
+              break
+          else
+            addOrderChange(linkToMove, orderedActiveLinks[i], orderChangesBetweenLinksToPerform)
+      when 'DOWN'
+        for i in [(startIndex + 1)...orderedActiveLinks.length]
+          if orderedActiveLinks[i] in linksToPassBy
+            removeOrderChange(linkToMove, orderedActiveLinks[i], orderChangesBetweenLinksToPerform)
+            linksToPassBy.splice(linksToPassBy.indexOf(orderedActiveLinks[i]), 1)
+            if linksToPassBy.length == 0
+              document.head.removeChild(linkToMove)
+              triggerEvent('page:after-link-removed', linkToMove)
+              orderedActiveLinks.splice(startIndex, 1)
+              targetIndex = i - 1
+
+              if targetIndex == orderedActiveLinks.length - 1
+                document.head.appendChild(linkToMove)
+                orderedActiveLinks.push(linkToMove)
+              else
+                document.head.insertBefore(linkToMove, orderedActiveLinks[targetIndex + 1])
+                orderedActiveLinks.splice(targetIndex + 1, 0, linkToMove)
+              triggerEvent('page:after-link-inserted', linkToMove)
+              break
+          else
+            addOrderChange(linkToMove, orderedActiveLinks[i], orderChangesBetweenLinksToPerform)
```

Comment:
Yeah I can abstract the bulk of the traversal and removal logic into a new function leaving the difference between them only in the injection of the link tag. 👍